remove global http client requirement

james-lawrence opened the pull request:
cleaned up version of #68 and #85 for #27. the package level scope isn't enough. we run a number of clients simultaneously and sometimes need to enable debugging of the http req/resp on a per client basis.

- allows for per instance http client configuration, useful for setting a client with a customized transport/timeouts etc.
- backwards compatible.

Commit:
remove global http client

dlsniper commented:
Sorry for the delay.
This is technically an API breakage and I'd rather not do this now.
I plan to tag this as a 0.9 at some point and then refactor everything out to have a better handling of the http.Client

dlsniper commented on misc.go:
```diff
@@ -16,8 +16,6 @@ import (
 	"time"
 )
 
-var HTTPClient = &http.Client{}
```
Removing this line causes an API break.

james-lawrence replied:
look again sorry I moved it to slack.go. sorry

james-lawrence commented on oauth.go:
```diff
@@ -45,7 +46,7 @@ func GetOAuthResponse(clientID, clientSecret, code, redirectURI string, debug bo
 		"redirect_uri":  {redirectURI},
 	}
 	response := &OAuthResponse{}
-	err = post("oauth.access", values, response, debug)
```
should probably change this to HTTPClient instead of http.DefaultClient

james-lawrence commented:
@dlsniper I disgree that it is an api breakage. I simply moved the HTTPClient to slack.go. misc.go felt like a random place to put it originally and I had trouble finding it.

I'm happy to rebase this again if you're willing to accept the change.

dlsniper commented:
Hi, I'll have to look at it again, let me come back to you at the end of the week please.

james-lawrence commented:
updated against latest master.

not a huge fan of the new customHTTPRequester adds in another global but it didn't change the PR much.
I did move the customHTTPRequester next to the HTTPClient in slack.go so they are near each other so the interaction is obvious from the code.

I also removed the getHTTPRequester() method as it is no longer necessary.

ernestoalejo commented:
This will help running this package in App Engine too. We need to pass a custom client with a context different for each request; we can't pass or initialize a global HTTP client in App Engine.

james-lawrence commented:
@dlsniper any update?

james-lawrence commented:
github isnt detecting my updates fixing the conflicts. closing and reopening.
